### DISCUSSION TITLE
Support parameters in IDataPortal<>.Execute, IDataPortal<>.ExecuteAsync

### DISCUSSION THREAD

**Discussion by hoangcuongvn** [MarimerLLC/csla#2923] [Ideas]
Actually `IDataPortal<>.Execute` and `IDataPortal<>.ExecuteAsync` does not support parameters. If I have to create a command class `VerifyCredentialsCommand`, it will be something like below:
```
public class VerifyCredentialsCommand : CommandBase<VerifyCredentialsCommand>
{
	public static readonly PropertyInfo<string> UserNameProperty = RegisterProperty<string>(x => x.UserName);
	public string UserName
	{
		get => ReadProperty(UserNameProperty);
		private set => LoadProperty(UserNameProperty, value);
	}

	public static readonly PropertyInfo<string> PasswordProperty = RegisterProperty<string>(x => x.Password);
	public string Password
	{
		get => ReadProperty(PasswordProperty);
		private set => LoadProperty(PasswordProperty, value);
	}

	public static readonly PropertyInfo<bool> OkProperty = RegisterProperty<bool>(x => x.Ok);
	public bool Ok
	{
		get => ReadProperty(OkProperty);
		private set => LoadProperty(OkProperty, value);
	}

	[Create]
	private void Create(string username, string password)
	{
		this.UserName = username;
		this.Password = password;
	}

	[Execute]
	private void Execute([Inject] DB db)
        {
           // Code here
        }
}
```
That means I have to create 2 properties `UserName`, `Password` and a `Create` method. 

If `IDataPortal<>.Execute`, `IDataPortal<>.ExecuteAsync` support parameters (E.g `T Execute(T obj, params object[] parameters)`) the class `VerifyCredentialsCommand` would become much more compact. It no longer needs the `UserName`, `Password` properties and `Create` method. It will be simple and compact like below
```
public class VerifyCredentialsCommand : CommandBase<VerifyCredentialsCommand>
{
	public static readonly PropertyInfo<bool> OkProperty = RegisterProperty<bool>(x => x.Ok);
	public bool Ok
	{
		get => ReadProperty(OkProperty);
		private set => LoadProperty(OkProperty, value);
	}

	[Execute]
	private void Execute(string username, string password, [Inject] DB db)
        {
           // Code here
        }
}
```
The code to execute the command looks like below
```
public bool VerifyCredentials(string username, string password)
{
    var dp = _context.GetRequiredService<IDataPortal<VerifyCredentialsCommand>>()
    var cmd = dp.Execute(dp.Create(), username, password);
    return cmd.Ok;
}
```

**Comment by rockfordlhotka** [comment]
This is by design. A command object is a bi-directional construct that flows from the logical client to the logical server with information, acts on that information while on the logical server, then returns to the logical client with information.

You can use `ReadOnlyBase` if you want to create a construct that sends criteria to the logical server, acts on that criteria, and returns to the logical client with information. (based on a fetch operation)

These are both types of "command", just two different ways to support the concept.

**Reply by hoangcuongvn** [reply]
How about a command that does not return a value? We cannot use `ReadOnlyBase` in this case. I mean we should not use "fetch" for a "not fetch" purpose.

**Reply by rockfordlhotka** [reply]
Most commands at least return a success/fail value. Even with `CommandBase`, the resulting object comes back to the caller.

Your point about "misusing" a fetch operation is valid though.

This is probably all lost in the mists of time, but CSLA didn't originally have an execute operation at all. Execute is really just a different name for "update" or "save" - in the data portal they all do the same thing.

Someone with great passion for the concept argued at length for the concept to be added. I remember making (and using for years) the joke that the data portal started with CRUD and then became CRUDE 😀

I did ultimately implement the execute operation, and if you walk through it in the data portal it largely flows through the process just like an update operation - there's just a bit of code on the server that understands that the original call was "execute" and so redirects the update operation to an execute method.

I think I am hearing two things.

1. There could be an execute that basically does the same thing as a fetch operation - take criteria, flow to the server, do work, and return the generated object with results
2. There could be some sort of one-way operation that takes criteria, flows to the server, and doesn't return anything at all

Is that an accurate summary of what you are suggesting?

---

Based on `CommandBase`:

```c#
  var cmd = await portal.CreateAsync(); // get command
  // set properties on cmd
  cmd = await portal.ExecuteAsync(); // execute command
  // use properties on cmd to see results
```

Based on `ReadOnlyBase`:

```c#
  var result = await portal.FetchAsync(1, 2, 3); // execute command
  // use properties on result to see results
```

New idea to avoid misusing fetch:

```c#
  var result = await portal.ExecuteAsync(1, 2, 3); // execute command
  // use properties on result to see results
```

New idea for one-way command:

```c#
  await portal.ExecuteOneWayAsync(1, 2, 3); // execute command
```

**Reply by hoangcuongvn** [reply]
Does the portal always need an instance of command when executing? I think `ExecuteOneWayAsync` is not necessary.
```
// If we need a returned value
var result = await portal.ExecuteAsync(cmd, 1, 2, 3);

// If we do not need a returned value. The command will throw an exception if failed.
await portal.ExecuteAsync(cmd, 1, 2, 3);
```

**Reply by rockfordlhotka** [reply]
The command object, in the execute with parameters case, _will_ flow back from the server to the client, yes. You could, of course, ignore the return value from the method, but the data portal will flow the data portal result from the server back to the client (which is how an exception would get back too).

**Reply by michaelcsikos** [reply]
So with `var result = await portal.ExecuteAsync(1, 2, 3);` we would no longer need to `Create` the command first? I like that.

**Reply by michaelcsikos** [reply]
This would also address [CommandBase requires [Create] method? #2297](https://github.com/MarimerLLC/csla/discussions/2297)

**Reply by rockfordlhotka** [reply]
Yes, for the execute with parameters scenario the domain object will be created on the server-side. It will literally work like an alias for fetch - but with a better name.

**Reply by hoangcuongvn** [reply]
I always love `Execute`, `ExecuteAsync` without having to create a command first 👍

**Comment by rockfordlhotka** [comment]
fwiw, this request is already on the backlog: https://github.com/MarimerLLC/csla/issues/1775

**Reply by hoangcuongvn** [reply]
Is there any news on this feature (Execute with parameters)? If you worry about breaking prior Csla's design, just take a small step by only adding optional parameters to the `IDataPortal.Execute` and `IDataPortal.ExecuteAsync`. This small change would help us much.
```
public interface IDataPortal<T>
{
    ...
    Task<T> ExecuteAsync(T command, params object[] parameters);
    T Execute(T obj, params object[] parameters);
}
```

**Reply by rockfordlhotka** [reply]
I hope to get it into v6. 1.
